src/Widgets/NotificationEntry.vala

```diff
@@ -186,6 +193,33 @@ public class Notifications.NotificationEntry : Gtk.ListBoxRow {
 
         grid.attach (body_label, 1, 1, 2);
 
+        if (notification.actions_with_label.length > 0) {
+            var action_area = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 6) {
+                margin_top = 12,
+                halign = Gtk.Align.END,
+                homogeneous = true
+            };
+
+            bool action_area_packed = false;
+            foreach (var action_name in notification.actions_with_label.get_keys ()) {
+                var button = new Gtk.Button.with_label (notification.actions_with_label[action_name]) {
+                    action_name = ACTION_PREFIX + action_name
+                };
+
+                button.clicked.connect (() => {
+                    activate ();
+                    clear (false);
+                });
+
+                action_area.pack_end (button);
+
+                if (!action_area_packed) {
+                    grid.attach (action_area, 0, 2, 3);
+                    action_area_packed = true;
+                }
```

Comment:
you don't need to do this inside the loop, the actions are already validated and it's known that we will pack the action_area.

Reply:
done